crates/uv-resolver/src/pins.rs

```diff
@@ -10,15 +10,14 @@ use crate::candidate_selector::Candidate;
 /// For example, given `Flask==3.0.0`, the [`FilePins`] would contain a mapping from `Flask` to
 /// `3.0.0` to the specific wheel or source distribution archive that was pinned for that version.
 #[derive(Clone, Debug, Default)]
-pub(crate) struct FilePins(FxHashMap<PackageName, FxHashMap<uv_pep440::Version, ResolvedDist>>);
+pub(crate) struct FilePins(FxHashMap<(PackageName, uv_pep440::Version), ResolvedDist>);
 
 impl FilePins {
     /// Pin a candidate package.
     pub(crate) fn insert(&mut self, candidate: &Candidate, dist: &CompatibleDist) {
-        self.0.entry(candidate.name().clone()).or_default().insert(
-            candidate.version().clone(),
-            dist.for_installation().to_owned(),
-        );
+        self.0
+            .entry((candidate.name().clone(), candidate.version().clone()))
+            .or_insert_with(|| dist.for_installation().to_owned());
```

Comment:
It's also important to only clone when we didn't insert the exact value yet in previous decision.